scripts/test_requirements.sh

```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+#
+# Check whether requirements*.in been changed without updating requirements*.txt
+#
+# We want to make sure that we are running `pip-compile` whenever we change our
+# Python requirements.
+#
+# In this repo the requirements files are generated from `*.in` files. This
+# script checks whether an in file has changed but the corresponding requirements
+# file has not.
+#
+# Returns 1 if `pip-compile` needs to be run, otherwise 0
+
+
+# Requirements files to check
+REQUIREMENTS_FILES=${@:-requirements.txt requirements-dev.txt}
+
+# Git reference to compare against
+BASE_COMMIT=${BASE_COMMIT:-master}
+HEAD_COMMIT=${HEAD_COMMIT:-HEAD}
+
+# Return 1 if file(s) has changed relative to BASE_COMMIT, otherwise 0
+test_unchanged() {
+  git diff --quiet $BASE_COMMIT $HEAD_COMMIT -- $@
+}
+
+# Return filename for requirements file
+requirements_infile() {
+  echo ${1%%.txt}.in
+}
+
+# Return 1 if requirement infile has chaged but not requirement file,
+# otherwise 0
+test_requirements_file() {
+  local requirements_file=$1
+  if [ ! -f $requirements_file ]
+  then
+    return 0
+  fi
+
+  local infile=$(requirements_infile $requirements_file)
+
+  test_unchanged $infile || (! test_unchanged $requirements_file)
+}
+
+status=0
+for requirements_file in $REQUIREMENTS_FILES
+do
+  if ! test_requirements_file $requirements_file
+  then
+    >&2 echo "$(requirements_infile $requirements_file) has changed since $BASE_COMMIT but $requirements_file has not"
+    status=1
+  fi
+done
+
+if [ $status -ne 0 ]
+then
+  # we want to allow changes to infiles that don't affect the requirements file
+  # the easiest way to do this is to run `pip-compile` and test the output
+  >&2 echo 'Running `make freeze-requirements`...'
+  make freeze-requirements > /dev/null 2>&1
```

Comment:
this is a bit of a loop: Makefile -> Bash -> Makefile, how about having on the makefile something like

```
freeze-requirements: virtualenv requirements-dev requirements.in
	./scripts/test_requirements.sh || ${VIRTUALENV_ROOT}/bin/pip-compile
	./scripts/test_requirements.sh dev || ${VIRTUALENV_ROOT}/bin/pip-compile requirements-dev.in
```
drop the test-requirements, and just update the Makefile dependencies accordingly

(you can also have an if block instead, if you prefer to have the test requirements run for both instead of per file)

Reply:
Oh!, I hadn't quite realised what you were doing here... although the same principle still applies.
Having a changed file that was not committed reinforces a bit my case for having a pre-commit hook too, I for one don't quite trust any checks that rely on humans. I need to do some thinking about how to pair this with dependabots though...